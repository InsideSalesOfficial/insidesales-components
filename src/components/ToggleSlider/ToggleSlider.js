import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import {
  colors,
  renderThemeKeyOrDefaultValue,
  renderThemeIfPresentOrDefault,
  ifThemeInPropsIsPresentUse
} from '../styles';

const SliderBase = styled.label`
  cursor: ${props => props.disabled ? 'not-allowed' : 'pointer'};
  display: inline-block;
  height: 14px;
  position: relative;
  width: 37px;
  ${props => ifThemeInPropsIsPresentUse({ props, value: props.disabled ? 'opacity: 0.6;' : '' })}
`;

const SliderInput = styled.input`
  display: none;
`;

function renderDefaultSliderKnobColors(props) {
  if (props.disabled) return colors.grayC;
  if (props.checked) return colors.green;
  return colors.lightGray;
}

function renderBrandedSliderKnobColors(props) {
  if (!props.checked) return props.theme.white40;
  return props.theme.brand01;
}

const SliderTrack = styled.div`
  background-color: ${props => {
    if (!props.checked) return renderThemeKeyOrDefaultValue({ props, key: 'white10', defaultValue: colors.grey });
    return renderThemeKeyOrDefaultValue({ props, key: 'brand03', defaultValue: colors.green70 });
  }};
  border-radius: 34px;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
  transition: .4s;

  &:before {
    background-color: ${props => ifThemeInPropsIsPresentUse({ props, value: renderBrandedSliderKnobColors(props), defaultValue: renderDefaultSliderKnobColors(props) })};
    border-radius: 50%;
    box-shadow: 0 1px 1px ${renderThemeIfPresentOrDefault({ key: 'black', defaultValue: colors.black })};
    content: "";
    height: 20px;
    left: 0px;
    position: absolute;
    top: -3px;
    transform: ${(props) => {
      if (props.checked) {
        return 'translateX(17px)';
      }

      return 'translateX(0)';
    }};
    transition: .4s;
    width: 20px;
  }
`;

class ToggleSlider extends React.PureComponent {
  render() {
    return (
      <SliderBase disabled={this.props.disabled}>
        <SliderInput
          className="pb-test__toggle-slider"
          type="checkbox"
          disabled={this.props.disabled}
          onChange={this.props.toggle}
          checked={this.props.checked} />
        <SliderTrack checked={this.props.checked} disabled={this.props.disabled} />
      </SliderBase>
    );
  }
}

ToggleSlider.defaultProps = {
  /** Function to execute when toggle is clicked */
  toggle: () => {},
  /** Boolean of whether the toggle is on */
  checked: false,
  disabled: false
};

ToggleSlider.propTypes = {
  checked: PropTypes.bool.isRequired,
  disabled: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired
};

export default ToggleSlider;
