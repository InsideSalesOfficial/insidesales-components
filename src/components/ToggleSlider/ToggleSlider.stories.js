import React from 'react';
import { storiesOf, action } from '@storybook/react';

import ToggleSlider from './';

class ToggleSliderWrapper extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: true
    };
  }
  render() {
    return <ToggleSlider checked={this.state.checked} toggle={() => {
      this.props.toggle();
      this.setState({ checked: !this.state.checked })
    }
    } />
  }
}


storiesOf('Forms', module)
  .addWithChapters(
  'ToggleSlider',
  {
    info: `
            Usage

            ~~~
            import React from 'react';
            import ToggleSlider from 'ui-components/ToggleSlider';
            ~~~
            `,
    chapters: [
      {
        sections: [
          {
            title: 'Example: Toggled On',
            subtitle: 'A slider toggled on',
            sectionFn: () => (
              <ToggleSlider
                toggle={action('toggle slider clicked')}
                checked={true} />
            )
          },
          {
            title: 'Example: Toggled Off',
            subtitle: 'A slider toggled off',
            sectionFn: () => (
              <ToggleSlider
                toggle={action('toggle slider clicked')}
                checked={false} />
            )
          },
          {
            title: 'Example: Wrapped in another component',
            subtitle: 'A toggle slider wrapped by another component that updates the checked prop on toggle',
            sectionFn: () => (
              <ToggleSliderWrapper toggle={action("toggled")} />
            )
          }
        ]
      }
    ]
  }
  );