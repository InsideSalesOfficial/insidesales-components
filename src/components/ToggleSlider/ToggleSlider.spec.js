import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import _ from 'lodash';

import ToggleSlider from './ToggleSlider';

describe('ToggleSlider', () => {
  it('toggle is called on click', () => {
    const onToggle = jest.fn();
    
    shallow(<ToggleSlider checked={false} toggle={onToggle} />);

    setTimeout(() => {
      expect(onToggle).toHaveBeenCalled();
    }, 0)
  });
});

describe('ToggleSlider Snapshots', () => {
  test('shows toggled on', () => {
    const tree = renderer.create(<ToggleSlider checked={true} toggle={_.noop} />);
    expect(tree).toMatchSnapshot();
  })
  test('shows toggled off', () => {
    const tree = renderer.create(<ToggleSlider checked={false} toggle={_.noop}/>);
    expect(tree).toMatchSnapshot();
  });
});