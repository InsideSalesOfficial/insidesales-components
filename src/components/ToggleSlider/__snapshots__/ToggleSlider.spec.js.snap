// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSlider Snapshots shows disabled 1`] = `
<label
  className="sc-bdVaJa kSRYIC"
  disabled={true}
>
  <input
    checked={false}
    className="pb-test__toggle-slider sc-bwzfXH eKeXgs"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    checked={false}
    className="sc-htpNat hnrsOU"
    disabled={true}
  />
</label>
`;

exports[`ToggleSlider Snapshots shows toggled off 1`] = `
<label
  className="sc-bdVaJa cMcrxJ"
  disabled={false}
>
  <input
    checked={false}
    className="pb-test__toggle-slider sc-bwzfXH eKeXgs"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    checked={false}
    className="sc-htpNat dwgooG"
    disabled={false}
  />
</label>
`;

exports[`ToggleSlider Snapshots shows toggled on 1`] = `
<label
  className="sc-bdVaJa cMcrxJ"
  disabled={false}
>
  <input
    checked={true}
    className="pb-test__toggle-slider sc-bwzfXH eKeXgs"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    checked={true}
    className="sc-htpNat dFyors"
    disabled={false}
  />
</label>
`;
